langchain/tools/base.py

```diff
@@ -1,19 +1,84 @@
 """Base implementation for tools or skills."""
 
-from abc import abstractmethod
-from typing import Any, Optional
+import inspect
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, Type, Union
 
-from pydantic import BaseModel, Extra, Field, validator
+from pydantic import BaseModel, Extra, Field, create_model, validator
 
 from langchain.callbacks import get_callback_manager
 from langchain.callbacks.base import BaseCallbackManager
 
 
-class BaseTool(BaseModel):
-    """Class responsible for defining a tool or skill for an LLM."""
+def create_args_schema_model_from_signature(run_func: Callable) -> Type[BaseModel]:
+    """Create a pydantic model type from a function's signature."""
+    signature_ = inspect.signature(run_func)
```

Comment:
This code is a bit hard to follow for me lol, do we really need to do this? Why can't we just just enforce that tools which take in `dict` params moving forward have to define their input schema as pydantic? Seems simpler than having to create a model on the fly